File: Design-Patterns/Strategy/CSharp/Character.cs
using Strategy.Attributes;
using Strategy.AttributeStrategies;

namespace Strategy
{
    public class Character
    {
        public IAttributes Attributes
        {
            get
            {
                Attributes.Attributes attributes = new Attributes.Attributes();
                attributes.Health = _attributeStrategy.Health;
                attributes.Mana = _attributeStrategy.Mana;
                attributes.Attack = _attributeStrategy.Attack;
                attributes.MagicalAttack = _attributeStrategy.MagicalAttack;
                attributes.Defense = _attributeStrategy.Defense;
                attributes.MagicalDefense = _attributeStrategy.MagicalDefense;

                return attributes;
            }
        }

        private AttributeStrategy _attributeStrategy;
        private readonly CharacterProperties _properties;

        public Character(CharacterProperties properties, AttributeStrategy attributeStrategy)
        {
            _properties = properties;
            _attributeStrategy = attributeStrategy;
            _attributeStrategy.Properites = properties;
        }

        public void SetAttributeStrategy<T>() where T : AttributeStrategy, new()
        {
            T strategy = new T
            {
                Properites = _properties
            };

            _attributeStrategy = strategy;
        }
    }
}

File: Design-Patterns/Strategy/Cpp/character.cpp
#include "character.h"

character::character(std::shared_ptr<character_properties> properties, std::shared_ptr<::attribute_strategy> attribute_strategy)
{
	properties_ = properties;
	attribute_strategy_ = attribute_strategy;
	attribute_strategy_->set_character_properties(properties);
}

void character::set_attribute_strategy(std::shared_ptr<attribute_strategy> attribute_strategy)
{
	attribute_strategy_ = attribute_strategy;
	attribute_strategy_->set_character_properties(properties_);
}

std::shared_ptr<i_attributes> character::get_attributes()
{
	std::shared_ptr<attributes> attributes = std::make_shared<::attributes>();
	
	attributes->set_health(attribute_strategy_->get_health());
	attributes->set_mana(attribute_strategy_->get_mana());
	attributes->set_attack(attribute_strategy_->get_attack());
	attributes->set_magical_attack(attribute_strategy_->get_magical_attack());
	attributes->set_defense(attribute_strategy_->get_defense());
	attributes->set_magical_defense(attribute_strategy_->get_magical_defense());

	return attributes;
}
